File: starters/rust-batch/src/application/jobs/db_init.rs
//! Database Initialization Job
//!
//! Initializes SQLite database and runs migrations.
//! Should be run once before using other database jobs.
//!
//! Usage:
//!   cargo run -- --job db-init

use std::fs;
use std::path::Path;
use std::sync::Arc;

use tracing::{debug, info};

use crate::application::job_factory::{create_job, JobDefinition, JobOptions};
use crate::application::job_types::JobResult;
use crate::di::AppContext;

/// Create the db-init job
pub fn db_init_job() -> JobDefinition {
    create_job(
        JobOptions::new("db-init", "Initialize SQLite database and run migrations"),
        |_ctx, app_ctx: Arc<AppContext>| async move {
            let migrations_dir = Path::new("db/migrations");

            info!(migrations_dir = ?migrations_dir, "Running database migrations");

            // Read migration files
            let files = if migrations_dir.exists() {
                let mut files: Vec<_> = fs::read_dir(migrations_dir)?
                    .filter_map(|entry| entry.ok())
                    .filter(|entry| {
                        entry
                            .path()
                            .extension()
                            .map(|ext| ext == "sql")
                            .unwrap_or(false)
                    })
                    .map(|entry| entry.path())
                    .collect();
                files.sort();
                files
            } else {
                return Ok(
                    JobResult::success("No migrations directory found").with_metric("applied", 0)
                );
            };

            if files.is_empty() {
                return Ok(JobResult::success("No migrations found").with_metric("applied", 0));
            }

            let mut applied_count = 0;

            let db = app_ctx.db.lock().map_err(|e| anyhow::anyhow!("{}", e))?;

            for file_path in files {
                let migration_name = file_path
                    .file_stem()
                    .and_then(|s| s.to_str())
                    .unwrap_or("unknown");

                // Check if migrations table exists
                let table_exists: bool = db
                    .query_row(
                        "SELECT COUNT(*) FROM sqlite_master WHERE type='table' AND name='migrations'",
                        [],
                        |row| row.get::<_, i32>(0),
                    )
                    .map(|count| count > 0)
                    .unwrap_or(false);

                if table_exists {
                    // Check if migration already applied
                    let existing: bool = db
                        .query_row(
                            "SELECT COUNT(*) FROM migrations WHERE name = ?1",
                            [migration_name],
                            |row| row.get::<_, i32>(0),
                        )
                        .map(|count| count > 0)
                        .unwrap_or(false);

                    if existing {
                        debug!(migration = %migration_name, "Skipping already applied migration");
                        continue;
                    }
                }

                // Read and execute SQL
                let sql = fs::read_to_string(&file_path)?;

                db.execute_batch(&sql)?;

                // Record migration (table should exist now after first migration)
                db.execute(
                    "INSERT INTO migrations (name) VALUES (?1)",
                    [migration_name],
                )?;

                info!(migration = %migration_name, "Applied migration");
                applied_count += 1;
            }

            let message = if applied_count > 0 {
                format!("Applied {} migration(s)", applied_count)
            } else {
                "All migrations already applied".to_string()
            };

            Ok(JobResult::success(message).with_metric("applied", applied_count))
        },
    )
}

File: optional/typescript-workflow-runner/src/application/jobs/database/db-init.ts
#!/usr/bin/env node
/**
 * Database Initialization Job
 *
 * Initializes SQLite database and runs migrations.
 * Should be run once before using other database jobs.
 *
 * Usage:
 *   npm run db-init
 *   tsx src/jobs/database/db-init.ts
 *
 * @module jobs/database/db-init
 */

import "../../../infra/config/yaml.js";
import { readdirSync, readFileSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";
import { createJobWithDeps } from "../../job-factory.js";
import { registerJob } from "../../../registry/job-registry.js";
import { runIfMain } from "../../../runner/workflow-runner.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

interface MigrationRow {
  name: string;
}

/**
 * Database initialization job
 *
 * Reads migration files from src/db/migrations/ and applies them
 * in order. Tracks applied migrations in the migrations table.
 */
export const dbInitJob = createJobWithDeps(
  {
    name: "db-init",
    description: "Initialize SQLite database and run migrations",
  },
  async (_ctx, logger, deps) => {
    const { db } = deps;
    const migrationsDir = join(__dirname, "../../../db/migrations");

    logger.info({ migrationsDir }, "Running database migrations");

    // Read migration files
    let files: string[];
    try {
      files = readdirSync(migrationsDir)
        .filter((f) => f.endsWith(".sql"))
        .sort();
    } catch {
      throw new Error(`Migrations directory not found: ${migrationsDir}`);
    }

    if (files.length === 0) {
      return {
        success: true,
        message: "No migrations found",
        metrics: { applied: 0 },
      };
    }

    let appliedCount = 0;

    for (const file of files) {
      const migrationName = file.replace(".sql", "");

      // Check if migrations table exists and migration already applied
      const tableExists = db
        .prepare(
          "SELECT name FROM sqlite_master WHERE type='table' AND name='migrations'"
        )
        .get();

      if (tableExists) {
        const existing = db
          .prepare("SELECT name FROM migrations WHERE name = ?")
          .get(migrationName) as MigrationRow | undefined;

        if (existing) {
          logger.debug(
            { migration: migrationName },
            "Skipping already applied migration"
          );
          continue;
        }
      }

      // Read and execute SQL
      const sql = readFileSync(join(migrationsDir, file), "utf-8");

      try {
        db.exec(sql);
        db.prepare("INSERT INTO migrations (name) VALUES (?)").run(
          migrationName
        );
        logger.info({ migration: migrationName }, "Applied migration");
        appliedCount++;
      } catch (error) {
        const message =
          error instanceof Error ? error.message : "Unknown error";
        throw new Error(`Migration failed: ${migrationName} - ${message}`);
      }
    }

    return {
      success: true,
      message:
        appliedCount > 0
          ? `Applied ${appliedCount} migration(s)`
          : "All migrations already applied",
      metrics: { applied: appliedCount },
    };
  }
);

registerJob(dbInitJob);
export const main = dbInitJob.handler;
runIfMain(main, import.meta.url);
